src/server.rs

```diff
@@ -326,59 +327,87 @@ impl Handler {
-            // debug!(cmd = format!("{:?}", cmd));
-            // ```
-            //
-            // `tracing` provides structured logging, so information is "logged"
-            // as key-value pairs.
-            debug!(?cmd);
-
-            // Perform the work needed to apply the command. This may mutate the
-            // database state as a result.
-            //
-            // The connection is passed into the apply function which allows the
-            // command to write response frames directly to the connection. In
-            // the case of pub/sub, multiple frames may be send back to the
-            // peer.
-            cmd.apply(&self.db, &mut self.connection, &mut self.shutdown)
-                .await?;
+            match self.process_frame().await? {
+                ControlFlow::Continue(..) => continue,
+                ControlFlow::Break(..) => return Ok(()),
+            }
         }
 
         Ok(())
     }
+
+    /// Process a single connection.
+    #[instrument(level = "debug", name = "Handler::process_frame", skip(self))]
+    async fn process_frame(&mut self) -> crate::Result<ControlFlow<(), ()>> {
+        // While reading a request frame, also listen for the shutdown
+        // signal.
+        let maybe_frame = tokio::select! {
+            res = self.connection.read_frame() => res?,
+            _ = self.shutdown.recv() => {
+                // If a shutdown signal is received, return from `run`.
+                // This will result in the task terminating.
+                return Ok(ControlFlow::Break(()));
+            }
+        };
+
+        // If `None` is returned from `read_frame()` then the peer closed
+        // the socket. There is no further work to do and the task can be
+        // terminated.
+        let frame = match maybe_frame {
+            Some(frame) => frame,
+            None => return Ok(ControlFlow::Break(())),
+        };
+
+        // Convert the redis frame into a command struct. This returns an
+        // error if the frame is not a valid redis command.
+        let cmd = match Command::from_frame(frame) {
+            Ok(cmd) => cmd,
+            Err(cause) => {
+                // The frame was malformed and could not be parsed. This is
+                // probably indicative of an issue with the client (as opposed
+                // to our server), so we (1) emit a warning...
+                //
+                // The syntax here is a shorthand provided by the `tracing`
+                // crate. It can be thought of as similar to:
+                //      warn! {
+                //          cause = format!("{}", cause),
+                //          "failed to parse command from frame"
+                //      };
+                // `tracing` provides structured logging, so information is
+                // "logged" as key-value pairs.
+                warn! {
+                    %cause,
+                    "failed to parse command from frame"
+                };
```

Comment:
```suggestion
                warn!(%cause, "failed to parse command from frame");
```
also, if `ParseError` implements `std::error::Error`, we may want to use that here instead of `Display`?